pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,355 @@
+	var filter *property.WaitFilter
+	recreateView := false
+	for {
+
+		if t.ctx.Err() == context.Canceled {
+			log.Info("listener canceled")
+			break
+		}
+
+		if t.getListView() == nil || recreateView {
+			log.Infof("listenToTaskUpdates -> recreateView -> createTasksListView")
+			listView, err := createTasksListView(ctx, t.virtualCenter.Client.Client, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				time.Sleep(waitForUpdatesRetry)
+				continue
+			}
+			t.setListView(t.ctx, listView)
+		}
+
+		if filter == nil || recreateView {
+			filter = getListViewWaitFilter(t.getListView())
+			recreateView = false
+		}
+		log.Infof("Starting listening for updates...")
+		pc := property.DefaultCollector(t.virtualCenter.Client.Client)
+		err := property.WaitForUpdates(t.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Infof("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Infof("Got update for object %v properties %v", update.Obj, prop)
+					if prop.Name == infoPropertyName {
+						go t.handleTaskUpdate(ctx, prop, false)
+					}
+				}
+			}
+
+			return t.shouldStopListening
+		})
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v, vimClient: %v", err, t.virtualCenter.Client.Client.UserAgent)
+			if soap.IsSoapFault(err) {
+				switch soap.ToSoapFault(err).VimFault().(type) {
+				case *types.ManagedObjectNotFound, types.ManagedObjectNotFound:
+					recreateView = true
+				case *types.NotAuthenticated, types.NotAuthenticated:
+					log.Info("observed NotAuthenticated error. should recreate view\n")
+					// todo: do we need to recreate view here?
+					recreateView = true
+					if err := t.virtualCenter.Connect(ctx); err != nil {
+						log.Errorf("failed to connect to vCenter. err: %v", err)
+					} else {
+						log.Infof("connection to vc successful")
+					}
+				}
+			}
+			// if any error is received from vpxd, we immediately return the fault for all the pending tasks in the map
+			// this is consistent with the current implementation
+			for taskDetails := range t.taskToChannelMap {
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						Task:  taskDetails.Reference,
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go t.handleTaskUpdate(ctx, prop, taskDetails.MarkedForRemoval)
+			}
+			time.Sleep(waitForUpdatesRetry)
+		}
+	}
+}
+
+// handleTaskUpdate is processes each task update in a separate goroutine
+func (t *TaskViewManagerImpl) handleTaskUpdate(ctx context.Context, prop types.PropertyChange, isMarkedForRemoval bool) {
+	log := logger.GetLogger(ctx)
+	log.Infof("handleTaskUpdate for prop: %+v", prop)
+	taskInfo := prop.Val.(types.TaskInfo)
+	if taskInfo.State == types.TaskInfoStateQueued || taskInfo.State == types.TaskInfoStateRunning {
```

Comment:
Will the taskInfo.State always be set?

Reply:
Yes. (https://vdc-repo.vmware.com/vmwb-repository/dcr-public/bd170cdc-e03a-4701-b468-bcfba05da6b2/946a76df-5276-46fa-a10d-06f0f7d4078e/doc/vim.TaskInfo.html)